oeps/architectural-decisions/oep-0065-frontend-composability.rst

```diff
@@ -0,0 +1,492 @@
+- Improves composability across MFEs/domains. We have no way to show more than one MFE on the same page today except by using iframes or by creating hard dependencies between MFEs at build-time by extracting 'shared' code into a new library, like `frontend-component-header <https://github.com/openedx/frontend-component-header>`_, `frontend-component-footer <https://github.com/openedx/frontend-component-footer>`_, or `frontend-lib-content-components <https://github.com/openedx/frontend-lib-content-components>`_, which increases our dependency maintenance burden significantly.
+- Improves runtime extensiblity by allowing us to configure where an MFE's code should be loaded from, rather than needing to build it in to an app. It dovetails nicely with the FPF by providing us with a seamless, performant, and flexible way of extending our frontends without needing to rebuild the host.
+
+Shared Dependencies
+===================
+
+Sharing dependencies between MFEs at runtime compliments runtime module loading.
+
+As noted above, MFEs have a number of dependencies which are common between them, but which aren't actually shared at runtime in any way. The capability to share these dependencies - such as ``react``, ``@openedx/paragon``, etc. - would mitigate a great deal of our `Consistency`_ and `User and Developer Experience`_ issues.
+
+We expect the following packages - which are used in the vast majority of MFEs today - could all be shared between MFEs.
+
+.. list-table::
+   :widths: 30 20 50
+
+   * - **Package**
+     - **Estimated Size**
+     - **Notes**
+   * - @edx/brand
+     - Variable
+     -
+   * - @edx/frontend-component-footer
+     - 88.1k
+     -
+   * - @edx/frontend-component-header
+     - 156.9k
+     -
+   * - @edx/frontend-platform
+     - 355.3k
+     -
+   * - @openedx/paragon
+     - ~950k
+     - bundlephobia.com threw an error on the Paragon build, but we expect it's the largest package here.
+   * - classnames
+     - 0.8k
+     -
+   * - core-js
+     - 241.1k
+     -
+   * - prop-types
+     - 0.9k
+     -
+   * - react
+     - 6.4k
+     -
+   * - react-dom
+     - 130.2k
+     -
+   * - react-redux
+     - 11.2k
+     -
+   * - react-router
+     - 58.9k
+     -
+   * - react-router-dom
+     - 77.1k
+     -
+   * - redux
+     - 3.7k
+     - Ultimately should be replaced with @reduxjs/toolkit. This document isn't making a judgment on the utility of redux, just noting that it is currently in use in the vast majority of MFEs.
+   * - regenerator-runtime
+     - 6.6k
+     -
+
+Total size: ~2,087.2k unzipped
+
+Shared dependency advantages
+----------------------------
+
+- Encourages runtime consistency across all types of dependencies. If our MFEs use the same version, we significantly reduce the cognitive load and context switching involved in working with multiple MFEs, cut down on visual inconsistencies at runtime, etc.
```

Comment:
The use of "encourages" instead of "enforces" here seems intentional. I know we don't want to dive into the "you can use different versions of a dependency on different MFEs within the same shell (assuming the dependencies aren't singletons) but you won't get the benefits of shared dependencies" rabbit hole here as that is covered quite well later in the OEP, but when first reading this section I see "***Encourages*** runtime consistency..." and "***If*** our MFEs use the same version..." and feel like I'm missing a piece of the puzzle.

I don't have any thoughts on how to best foreshadow that (hopefully someone else does?)

Reply:
"When MFEs use the same version of a given dependency (runtime consistency), we see many benefits: significant reduction of developer cognitive load and context switching involved in working with multiple MFEs, fewer visual inconsistencies at runtime, and more. However, it will still be possible for MFEs to choose to use a different version of a dependency, which will be discussed in more detail later in this section."

How does that sound @brian-smith-tcril ?

Reply:
That sounds great! We probably don't even need the the part with "runtime consistency" in parens there.

Reply:
I like it - I replaced the last sentence with a link down to the "Approach" section which gets into the deets of what shared dependencies really means, and how it works (and what it takes to _enforce_ it. 😁).